In [330]:
import pandas as pd
import numpy as np

In [331]:
data=pd.read_csv('Iphone-11_Review.csv',index_col=0)

In [332]:
data.head()

name                                             titles  rating  \
0      Abhishek                                   iPhone 11 review       2   
1      malleshk                                  2 years completed       4   
2            PR                                  Beast with a limp       5   
3        stella  The best iPhone - perfect combination of featu...       4   
4  Gurlal Singh                                               good       5   

                                            comments  
0  iPhone11 Review A person who uses every major ...  
1  Battery lifePhone performanceCalling feature/q...  
2  Been almost 10 months I have been using the iP...  
3  This is my first iPhone. After being an Androi...  
4                                       good product

In [333]:
data.duplicated().sum()

28

In [334]:
data.drop_duplicates(inplace=True)

In [335]:
data.isna().sum()

name         0
titles       1
rating       0
comments    75
dtype: int64

In [336]:
data.dropna(inplace=True)

In [337]:
data.reset_index(inplace=True)

In [338]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3726 entries, 0 to 3725
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   index     3726 non-null   int64 
 1   name      3726 non-null   object
 2   titles    3726 non-null   object
 3   rating    3726 non-null   int64 
 4   comments  3726 non-null   object
dtypes: int64(2), object(3)
memory usage: 145.7+ KB


In [339]:
data.isna().sum()

index       0
name        0
titles      0
rating      0
comments    0
dtype: int64

In [340]:
data.duplicated().sum()

0

In [341]:
data1=data.loc[:,['comments']]

In [342]:
data1

comments
0     iPhone11 Review A person who uses every major ...
1     Battery lifePhone performanceCalling feature/q...
2     Been almost 10 months I have been using the iP...
3     This is my first iPhone. After being an Androi...
4                                          good product
...                                                 ...
3721                                      As expected !
3722                  Awsome i loved it, liked soo much
3723                Very good quality love it iphone 11
3724                                      Awesome phone
3725                                          very good

[3726 rows x 1 columns]

In [343]:
# as you can there are still number there
import re

In [344]:
def remove_pattern(text,pattern):
    x=re.findall(pattern,str(text))
    for i in x:
        text=re.sub(i,'',str(text))
    return text

In [345]:
data1['cleaned']=data1['comments'].apply(lambda x: remove_pattern(x,'[0-9]*'))
data1

comments  \
0     iPhone11 Review A person who uses every major ...   
1     Battery lifePhone performanceCalling feature/q...   
2     Been almost 10 months I have been using the iP...   
3     This is my first iPhone. After being an Androi...   
4                                          good product   
...                                                 ...   
3721                                      As expected !   
3722                  Awsome i loved it, liked soo much   
3723                Very good quality love it iphone 11   
3724                                      Awesome phone   
3725                                          very good   

                                                cleaned  
0     iPhone Review A person who uses every major ap...  
1     Battery lifePhone performanceCalling feature/q...  
2     Been almost  months I have been using the iPho...  
3     This is my first iPhone. After being an Androi...  
4                                          good product  
...                                                 ...  
3721                                      As expected !  
3722                  Awsome i loved it, liked soo much  
3723                  Very good quality love it iphone   
3724                                      Awesome phone  
3725                                          very good  

[3726 rows x 2 columns]

In [346]:
from string import punctuation
from nltk.corpus import stopwords

In [347]:
stop_words=list(stopwords.words('english'))
punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

In [348]:
from nltk import word_tokenize

In [349]:
data1['tokenized']=[word_tokenize(x) for x in data1['cleaned']]
data1

comments  \
0     iPhone11 Review A person who uses every major ...   
1     Battery lifePhone performanceCalling feature/q...   
2     Been almost 10 months I have been using the iP...   
3     This is my first iPhone. After being an Androi...   
4                                          good product   
...                                                 ...   
3721                                      As expected !   
3722                  Awsome i loved it, liked soo much   
3723                Very good quality love it iphone 11   
3724                                      Awesome phone   
3725                                          very good   

                                                cleaned  \
0     iPhone Review A person who uses every major ap...   
1     Battery lifePhone performanceCalling feature/q...   
2     Been almost  months I have been using the iPho...   
3     This is my first iPhone. After being an Androi...   
4                                          good product   
...                                                 ...   
3721                                      As expected !   
3722                  Awsome i loved it, liked soo much   
3723                  Very good quality love it iphone    
3724                                      Awesome phone   
3725                                          very good   

                                              tokenized  
0     [iPhone, Review, A, person, who, uses, every, ...  
1     [Battery, lifePhone, performanceCalling, featu...  
2     [Been, almost, months, I, have, been, using, t...  
3     [This, is, my, first, iPhone, ., After, being,...  
4                                       [good, product]  
...                                                 ...  
3721                                  [As, expected, !]  
3722        [Awsome, i, loved, it, ,, liked, soo, much]  
3723            [Very, good, quality, love, it, iphone]  
3724                                   [Awesome, phone]  
3725                                       [very, good]  

[3726 rows x 3 columns]

In [350]:
data1['tokenized']=data1['tokenized'].apply(lambda x: [word for word in x if word not in stop_words])

In [351]:
data1.head()

comments  \
0  iPhone11 Review A person who uses every major ...   
1  Battery lifePhone performanceCalling feature/q...   
2  Been almost 10 months I have been using the iP...   
3  This is my first iPhone. After being an Androi...   
4                                       good product   

                                             cleaned  \
0  iPhone Review A person who uses every major ap...   
1  Battery lifePhone performanceCalling feature/q...   
2  Been almost  months I have been using the iPho...   
3  This is my first iPhone. After being an Androi...   
4                                       good product   

                                           tokenized  
0  [iPhone, Review, A, person, uses, every, major...  
1  [Battery, lifePhone, performanceCalling, featu...  
2  [Been, almost, months, I, using, iPhone, GB, ....  
3  [This, first, iPhone, ., After, Android, enthu...  
4                                    [good, product]

In [352]:
data1

comments  \
0     iPhone11 Review A person who uses every major ...   
1     Battery lifePhone performanceCalling feature/q...   
2     Been almost 10 months I have been using the iP...   
3     This is my first iPhone. After being an Androi...   
4                                          good product   
...                                                 ...   
3721                                      As expected !   
3722                  Awsome i loved it, liked soo much   
3723                Very good quality love it iphone 11   
3724                                      Awesome phone   
3725                                          very good   

                                                cleaned  \
0     iPhone Review A person who uses every major ap...   
1     Battery lifePhone performanceCalling feature/q...   
2     Been almost  months I have been using the iPho...   
3     This is my first iPhone. After being an Androi...   
4                                          good product   
...                                                 ...   
3721                                      As expected !   
3722                  Awsome i loved it, liked soo much   
3723                  Very good quality love it iphone    
3724                                      Awesome phone   
3725                                          very good   

                                              tokenized  
0     [iPhone, Review, A, person, uses, every, major...  
1     [Battery, lifePhone, performanceCalling, featu...  
2     [Been, almost, months, I, using, iPhone, GB, ....  
3     [This, first, iPhone, ., After, Android, enthu...  
4                                       [good, product]  
...                                                 ...  
3721                                  [As, expected, !]  
3722               [Awsome, loved, ,, liked, soo, much]  
3723                [Very, good, quality, love, iphone]  
3724                                   [Awesome, phone]  
3725                                             [good]  

[3726 rows x 3 columns]

In [353]:
from nltk import PorterStemmer

In [354]:
stemm=PorterStemmer()

In [355]:
data1['tokenized']=data1['tokenized'].apply(lambda x:[stemm.stem(i) for i in x])

In [356]:
data1.head()

comments  \
0  iPhone11 Review A person who uses every major ...   
1  Battery lifePhone performanceCalling feature/q...   
2  Been almost 10 months I have been using the iP...   
3  This is my first iPhone. After being an Androi...   
4                                       good product   

                                             cleaned  \
0  iPhone Review A person who uses every major ap...   
1  Battery lifePhone performanceCalling feature/q...   
2  Been almost  months I have been using the iPho...   
3  This is my first iPhone. After being an Androi...   
4                                       good product   

                                           tokenized  
0  [iphon, review, a, person, use, everi, major, ...  
1  [batteri, lifephon, performancecal, feature/qu...  
2  [been, almost, month, i, use, iphon, gb, ., ap...  
3  [thi, first, iphon, ., after, android, enthusi...  
4                                    [good, product]

In [357]:
data1['tokenized']=data1['tokenized'].apply(lambda x:' '.join(x))

In [358]:
data1

comments  \
0     iPhone11 Review A person who uses every major ...   
1     Battery lifePhone performanceCalling feature/q...   
2     Been almost 10 months I have been using the iP...   
3     This is my first iPhone. After being an Androi...   
4                                          good product   
...                                                 ...   
3721                                      As expected !   
3722                  Awsome i loved it, liked soo much   
3723                Very good quality love it iphone 11   
3724                                      Awesome phone   
3725                                          very good   

                                                cleaned  \
0     iPhone Review A person who uses every major ap...   
1     Battery lifePhone performanceCalling feature/q...   
2     Been almost  months I have been using the iPho...   
3     This is my first iPhone. After being an Androi...   
4                                          good product   
...                                                 ...   
3721                                      As expected !   
3722                  Awsome i loved it, liked soo much   
3723                  Very good quality love it iphone    
3724                                      Awesome phone   
3725                                          very good   

                                              tokenized  
0     iphon review a person use everi major applic p...  
1     batteri lifephon performancecal feature/qu suberb  
2     been almost month i use iphon gb . appl known ...  
3     thi first iphon . after android enthusiast loy...  
4                                          good product  
...                                                 ...  
3721                                        as expect !  
3722                         awsom love , like soo much  
3723                       veri good qualiti love iphon  
3724                                       awesom phone  
3725                                               good  

[3726 rows x 3 columns]

In [359]:
data2=data1.loc[:,['tokenized']]

In [360]:
data2

tokenized
0     iphon review a person use everi major applic p...
1     batteri lifephon performancecal feature/qu suberb
2     been almost month i use iphon gb . appl known ...
3     thi first iphon . after android enthusiast loy...
4                                          good product
...                                                 ...
3721                                        as expect !
3722                         awsom love , like soo much
3723                       veri good qualiti love iphon
3724                                       awesom phone
3725                                               good

[3726 rows x 1 columns]

In [361]:
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

In [362]:
analyser=SentimentIntensityAnalyzer()

In [363]:
def sentiment_score(text):
    analyser=SentimentIntensityAnalyzer()
    senti_dict=analyser.polarity_scores(text)
    senti=[]
    for i in senti_dict:
        if senti_dict['compound']>0.5:
            senti.append(1)
        elif senti_dict['compound']<=-0.5:
            senti.append(2)
        else :
            senti.append(0)
    
    return senti[0]

In [364]:
data3=data2

In [365]:
data3

tokenized
0     iphon review a person use everi major applic p...
1     batteri lifephon performancecal feature/qu suberb
2     been almost month i use iphon gb . appl known ...
3     thi first iphon . after android enthusiast loy...
4                                          good product
...                                                 ...
3721                                        as expect !
3722                         awsom love , like soo much
3723                       veri good qualiti love iphon
3724                                       awesom phone
3725                                               good

[3726 rows x 1 columns]

In [366]:
data3['target']=data3['tokenized'].apply(lambda x: sentiment_score(x))

In [367]:
data3

tokenized  target
0     iphon review a person use everi major applic p...       1
1     batteri lifephon performancecal feature/qu suberb       0
2     been almost month i use iphon gb . appl known ...       1
3     thi first iphon . after android enthusiast loy...       1
4                                          good product       0
...                                                 ...     ...
3721                                        as expect !       0
3722                         awsom love , like soo much       1
3723                       veri good qualiti love iphon       1
3724                                       awesom phone       0
3725                                               good       0

[3726 rows x 2 columns]

In [368]:
data3[data3['target']==2]

tokenized  target
110   the camera qualiti appl phone get worst worst ...       2
124   it 's great product price i bought . use vario...       2
125   to honest , hesit order read neg comment . it ...       2
130   i first time iphon user . i bought iphon last ...       2
138   if switch android follow thing frustrat appl :...       2
...                                                 ...     ...
3351  veri bulki . so bulki finger hurt use phone.a ...       2
3428  amaz phone appl ..... unbeliev batteri back-up...       2
3603  the media could load . touch sensit poor . pho...       2
3619           bad camera qualiti phone heat kept charg       2
3667  receiv empti packag waranti card user manual ....       2

[77 rows x 2 columns]

In [369]:
data3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3726 entries, 0 to 3725
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   tokenized  3726 non-null   object
 1   target     3726 non-null   int64 
dtypes: int64(1), object(1)
memory usage: 58.3+ KB


In [370]:
data3['target'].value_counts()

0    1858
1    1791
2      77
Name: target, dtype: int64

### now we using CountVector

In [371]:
from sklearn.feature_extraction.text import CountVectorizer

In [372]:
cv=CountVectorizer()
cv_text=cv.fit_transform(data3['tokenized'])

###  Logistic Regression

In [373]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from imblearn.over_sampling import SMOTE

In [374]:
lr=LogisticRegression()
sm=SMOTE()

In [375]:
x=cv_text
y=data3['target']

In [376]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=.33,random_state=42)

In [377]:
x_train,y_train=sm.fit_resample(x_train,y_train)

In [378]:
y_train.value_counts()

1    1270
0    1270
2    1270
Name: target, dtype: int64

In [379]:
model_log=lr.fit(x_train,y_train)

In [380]:
print('accuracy :',model_log.score(x_train,y_train)*100)

accuracy : 96.27296587926509


In [381]:
print('accuracy :',model_log.score(x_test,y_test)*100)

accuracy : 77.72357723577235


In [382]:
y_pred=model_log.predict(x_test)

In [383]:
print(classification_report(y_pred,y_test))

              precision    recall  f1-score   support

           0       0.75      0.80      0.77       548
           1       0.83      0.90      0.86       566
           2       0.21      0.04      0.07       116

    accuracy                           0.78      1230
   macro avg       0.59      0.58      0.57      1230
weighted avg       0.73      0.78      0.75      1230



###  Naive Bayes

In [384]:
from sklearn.naive_bayes import MultinomialNB

In [385]:
nb=MultinomialNB()
model_2=nb.fit(x_train,y_train)

In [386]:
print('accuracy :',model_2.score(x_train,y_train)*100)

accuracy : 75.64304461942257


In [387]:
print('accuracy :',model_2.score(x_test,y_test)*100)

accuracy : 68.78048780487805


In [388]:
y_pred=model_2.predict(x_test)

In [389]:
print(classification_report(y_pred,y_test))

              precision    recall  f1-score   support

           0       0.53      0.76      0.63       410
           1       0.86      0.69      0.76       768
           2       0.17      0.08      0.11        52

    accuracy                           0.69      1230
   macro avg       0.52      0.51      0.50      1230
weighted avg       0.72      0.69      0.69      1230



### Decision Tree

In [390]:
from sklearn.tree import DecisionTreeClassifier

In [391]:
dt=DecisionTreeClassifier()
model_3=dt.fit(x_train,y_train)

In [392]:
print('accuracy :',model_3.score(x_train,y_train)*100)

accuracy : 99.26509186351706


In [393]:
print('accuracy :',model_3.score(x_test,y_test)*100)

accuracy : 75.28455284552845


In [394]:
y_pred=model_3.predict(x_test)

In [395]:
print(classification_report(y_pred,y_test))

              precision    recall  f1-score   support

           0       0.76      0.75      0.75       593
           1       0.77      0.88      0.82       539
           2       0.25      0.06      0.10        98

    accuracy                           0.75      1230
   macro avg       0.59      0.56      0.56      1230
weighted avg       0.72      0.75      0.73      1230



### Now we are using TF-IDF

In [396]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [397]:
tf=TfidfVectorizer()
tf_text=tf.fit_transform(data3['tokenized'])

In [398]:
x=tf_text
y=data3['target']

In [399]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=.33,random_state=42)
x_train,y_train=sm.fit_resample(x_train,y_train)
y_train.value_counts()

1    1270
0    1270
2    1270
Name: target, dtype: int64

### Logistic regression

In [414]:
lr=LogisticRegression()
tf_model_1=lr.fit(x_train,y_train)

In [415]:
print('accuracy :',tf_model_1.score(x_train,y_train)*100)

accuracy : 94.33070866141732


In [419]:
print('accuracy :',tf_model_1.score(x_test,y_test)*100)

accuracy : 81.70731707317073


In [420]:
y_pred=tf_model_1.predict(x_test)

In [421]:
print(classification_report(y_pred,y_test)) #81

              precision    recall  f1-score   support

           0       0.86      0.78      0.82       649
           1       0.79      0.88      0.83       560
           2       0.29      0.33      0.31        21

    accuracy                           0.82      1230
   macro avg       0.65      0.66      0.65      1230
weighted avg       0.82      0.82      0.82      1230



### Naive bayes

In [422]:
tf_model_2=nb.fit(x_train,y_train)

In [423]:
print('accuracy :',tf_model_2.score(x_train,y_train)*100)

accuracy : 87.71653543307086


In [424]:
print('accuracy :',tf_model_2.score(x_test,y_test)*100)

accuracy : 67.96747967479675


In [425]:
y_pred=tf_model_2.predict(x_test)

In [426]:
print(classification_report(y_pred,y_test))

              precision    recall  f1-score   support

           0       0.51      0.79      0.62       376
           1       0.85      0.70      0.77       757
           2       0.46      0.11      0.18        97

    accuracy                           0.68      1230
   macro avg       0.61      0.53      0.52      1230
weighted avg       0.72      0.68      0.68      1230



### Decision Tree

In [428]:
dt=DecisionTreeClassifier()
tf_model_3=dt.fit(x_train,y_train)
print('accuracy :',tf_model_3.score(x_train,y_train)*100)

accuracy : 99.86876640419948


In [429]:
print('accuracy :',tf_model_3.score(x_test,y_test)*100)

accuracy : 78.86178861788618


In [430]:
y_pred=tf_model_3.predict(x_test)

In [431]:
print(classification_report(y_pred,y_test))

              precision    recall  f1-score   support

           0       0.84      0.77      0.80       646
           1       0.76      0.84      0.80       556
           2       0.25      0.21      0.23        28

    accuracy                           0.79      1230
   macro avg       0.62      0.61      0.61      1230
weighted avg       0.79      0.79      0.79      1230



## as we can see in TF-IDF if we use Logistic Rgeression we got 82% accuracy in test data so if i use other algorithms. then may get better model with better accuracy compare to Logistic Regression